src/nodeMatchers.ts

```diff
@@ -1,200 +1,35 @@
-import { SyntaxNode, Point } from "web-tree-sitter";
-import { Position, Range, Selection, TextEditor } from "vscode";
-import { SelectionWithContext, NodeMatcher } from "./Types";
-
-export function hasType(...typeNames: string[]): NodeMatcher {
-  return (editor: TextEditor, node: SyntaxNode) =>
-    typeNames.includes(node.type) ? simpleSelectionExtractor(node) : null;
-}
-
-export function childNodeMatcher(
-  getMatchingChildNode: (node: SyntaxNode) => SyntaxNode | null,
-  extractor: (node: SyntaxNode) => SelectionWithContext
+import { SyntaxNode } from "web-tree-sitter";
+import { TextEditor } from "vscode";
+import { NodeMatcher, NodeFinder, SelectionExtractor } from "./Types";
+import { simpleSelectionExtractor } from "./nodeSelectors";
+import { findNodeOfType } from "./nodeFinders";
+
+export function matcher(
+  finder: NodeFinder,
+  selector: SelectionExtractor = simpleSelectionExtractor
 ): NodeMatcher {
-  return (editor: TextEditor, node: SyntaxNode) => {
-    const returnNode = getMatchingChildNode(node);
-
-    if (returnNode == null) {
-      return null;
-    }
-
-    return extractor(returnNode);
+  return function (editor: TextEditor, node: SyntaxNode) {
+    const targetNode = finder(node);
+    return targetNode ? selector(editor, targetNode) : null;
   };
 }
 
-export function getNodeWithLeadingDelimiter(
-  node: SyntaxNode
-): SelectionWithContext {
-  const leadingDelimiterToken = node.previousSibling!;
-
-  const leadingDelimiterRange = makeRange(
-    leadingDelimiterToken.startPosition,
-    node.startPosition
-  );
-
-  return {
-    ...simpleSelectionExtractor(node),
-    context: {
-      leadingDelimiterRange,
-    },
-  };
-}
-
-export const notSupported: NodeMatcher = (
-  editor: TextEditor,
-  node: SyntaxNode
-) => {
-  throw new Error("Node type not supported");
-};
-
-function getNextNonDelimiterNode(
-  startNode: SyntaxNode,
-  isDelimiterNode: (node: SyntaxNode) => boolean
-): SyntaxNode | null {
-  var node = startNode.nextSibling;
-
-  while (node != null) {
-    if (!isDelimiterNode(node)) {
-      return node;
-    }
-
-    node = node.nextSibling;
-  }
-
-  return node;
-}
-
-function getPreviousNonDelimiterNode(
-  startNode: SyntaxNode,
-  isDelimiterNode: (node: SyntaxNode) => boolean
-): SyntaxNode | null {
-  var node = startNode.previousSibling;
-
-  while (node != null) {
-    if (!isDelimiterNode(node)) {
-      return node;
-    }
-
-    node = node.previousSibling;
-  }
-
-  return node;
-}
-
-export function delimitedMatcher(
-  nodeMatches: (node: SyntaxNode) => boolean,
-  isDelimiterNode: (node: SyntaxNode) => boolean,
-  defaultDelimiter: string
+export function composedMatcher(
```

Comment:
hmm i tend to like variadic signatures for this type of thing, so you can say eg `composedMatcher(findNodeOfType("pair"), getKeyNode)` (notice no square brackets).  But obv that doesn't work with this selector param.  Not sure what the right answer is here.  Prob fine as you have it

Reply:
Yeah I would have preferred that too.

I think maybe it can actually be done ([see this PR for examples](https://github.com/microsoft/TypeScript/pull/39094)), but when I tried it seemed to break type inference within the function so I backed off. I did not spend much time looking into this though.

It would look something like this (?):

```ts
export function composedMatcher(
    ...args: [...NodeFinder[], SelectionExtractor]
)
```

Reply:
Hmm I think that's potentially more confusing.  Maybe let's just leave it